### PR TITLE
added login to demo server

### PR DESCRIPTION
**Pull request by austincunningham**
## Motivation
The Keycloak authentication mechanism works based on a redirect to the Auth server for authentication. We need to mimic how this works in passportJS. 

## Description
Implement the following in the demo app server, 
We should implement the login redirect flow.
We should implement the login page with raincatcher styling.
We should provide better handling of the access denied message being returned.

## Progress
- [x] login redirect flow
- [x] login page with raincatcher styling

## Additional Notes
screenshot from 2017-06-28 14-41-33

npm run start
http://localhost:3000/login

### CODE REVIEW

File: demo/server/package.json [hunk 1]
```diff
@@ -46,9 +46,11 @@
     "typescript": "^2.3.4"
   },
   "dependencies": {
+    "@types/express-handlebars": "0.0.29",
```
Comment (wtrocki) [inline demo/server/package.json]: That's dev dependency used for builds.

File: demo/server/src/app.ts [hunk 2]
```diff
@@ -28,6 +29,9 @@ app.use((req: express.Request, res: express.Response, next) => {
   next(err);
 });
 
+app.engine('hbs', expressHbs());
```
Comment (wtrocki) [inline demo/server/src/app.ts]: Good to wrap this in method: `setupCustomLoginPageRenderer()` etc.

File: demo/server/src/routes/index.ts [hunk 3]
```diff
@@ -8,4 +8,15 @@ router.get('/', (req: express.Request, res: express.Response) => {
   res.json(api);
 });
 
+router.get('/login', (req: express.Request, res: express.Response) => {
```
Comment (wtrocki) [inline demo/server/src/routes/index.ts]: Best to have this as security routes (separate file) so it can be reused.

**Review by wtrocki — DISMISSED** [review]

File: demo/server/src/app.ts (hunk 2 above)
Comment (austincunningham) [reply demo/server/src/app.ts]: done

File: demo/server/src/routes/index.ts (hunk 3 above)
Comment (austincunningham) [reply demo/server/src/routes/index.ts]: done

File: client/datasync-client/dist/src/index.js [hunk 4]
```diff
@@ -0,0 +1,7 @@
+"use strict";
```
Comment (wtrocki) [inline client/datasync-client/dist/src/index.js]: Remove all js files changes from datasync.

File: demo/server/src/securityRoutes/index.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+'use strict';
```
Comment (wtrocki) [inline demo/server/src/securityRoutes/index.ts]: why create additional folder. Let's just do routes/security.

File: client/datasync-client/dist/src/index.js (hunk 4 above)
Comment (austincunningham) [reply client/datasync-client/dist/src/index.js]: npm run clean

File: demo/server/src/securityRoutes/index.ts (hunk 5 above)
Comment (austincunningham) [reply demo/server/src/securityRoutes/index.ts]: done

File: client/datasync-client/dist/src/index.js (hunk 4 above)
Comment (wtrocki) [reply client/datasync-client/dist/src/index.js]: 😆 I wanted to say that we should remove this file and all js files in this PR. 
Run `npm run clean` on your machine for that :)

File: client/datasync-client/dist/src/index.js.map [hunk 6]
```diff
@@ -0,0 +1 @@
+{"version":3,"file":"index.js","sourceRoot":"","sources":["../../src/index.ts"],"names":[],"mappings":";;AAAA,iCAAmC;AAG1B,oBAAI;AACb,uBAAuB;AACvB,kBAAe,IAAI,CAAC"}
```
Comment (wtrocki) [inline client/datasync-client/dist/src/index.js.map]: Same as above!

File: demo/server/src/app.ts [hunk 7]
```diff
@@ -17,17 +19,23 @@ app.use(logger(config.morganOptions));
 app.use(bodyParser.json());
 app.use(bodyParser.urlencoded({ extended: false }));
 app.use(cookieParser());
-app.use(express.static(path.join(__dirname, 'public')));
+app.use(express.static(path.join(__dirname, '../public')));
 app.use(favicon(path.join(__dirname, '../public', 'favicon.ico')));
 
-app.use('/', index);
+app.use('/', index, security);
```
Comment (wtrocki) [inline demo/server/src/app.ts:24]: Is that valid?

File: demo/server/src/app.ts [hunk 8]
```diff
@@ -17,17 +19,23 @@ app.use(logger(config.morganOptions));
...
   const err: any = new Error('Not Found');
   err.status = 404;
   next(err);
 });
 
+function setupCustomLoginPageRender() {
+  app.engine('hbs', expressHbs());
+  app.set('view engine', 'hbs');
+}
+setupCustomLoginPageRender();
```
Comment (wtrocki) [inline demo/server/src/app.ts]: This looks strange. Idea was to pull out login related code from entire file. 
This way it's even worse than before.

File: demo/server/src/app.ts (hunk 7 above)
Comment (austincunningham) [reply demo/server/src/app.ts:24]: both routes resolve so I guess so

File: client/datasync-client/dist/src/index.js.map (hunk 6 above)
Comment (austincunningham) [reply client/datasync-client/dist/src/index.js.map]: removed now

File: client/datasync-client/dist/src/index.js (hunk 4 above)
Comment (austincunningham) [reply client/datasync-client/dist/src/index.js]: npm run clean doesn't remove files fro dist directory , removed manually

File: demo/server/src/app.ts [hunk 9]
```diff
@@ -17,17 +19,23 @@ app.use(logger(config.morganOptions));
...
-app.use('/', index);
+app.use('/', index, security);
 
 app.use((req: express.Request, res: express.Response, next) => {
   const err: any = new Error('Not Found');
   err.status = 404;
   next(err);
 });
 
+function setupCustomLoginPageRender() {
```
Comment (wtrocki) [inline demo/server/src/app.ts]: This will be moved to passport integration code.

**Review by wtrocki — DISMISSED** [review]
Great job!

**Review by wtrocki — APPROVED** [review]

File: demo/server/views/login.hbs [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+              <div>
+                <div id="form-options">
+                  <div>
+                  </div>
+                </div>
+
+                <div id="form-buttons">
+                  <div >
+                    <input name="login" id="kc-login" type="submit" value="Submit"/>
```
Comment (JameelB) [inline demo/server/views/login.hbs:59]: maybe `rc-login` instead of `kc`?

**Review by JameelB — APPROVED** [review]

**Pull request merged**